File: csharp/Primrose.NETStandard2/Row.cs
using System;
using System.Globalization;

namespace Primrose
{

    public class Row
    {

        public static Row emptyRow(int startStringIndex, int startTokenIndex, int lineNumber)
        {
            return new Row(string.Empty, Array.Empty<Token>(), startStringIndex, startTokenIndex, lineNumber);
        }

        public string text { get; }
        public int startStringIndex { get; set; }
        public int stringLength => text.Length;
        public int endStringIndex => startStringIndex + stringLength;
        public Token[] tokens { get; }
        public int startTokenIndex { get; set; }
        public int numTokens => tokens.Length;
        public int endTokenIndex => startTokenIndex + numTokens;
        public int lineNumber { get; set; }

        private readonly int[] leftCorrections;
        private readonly int[] rightCorrections;

        public Row(string txt, Token[] tokens, int startStringIndex, int startTokenIndex, int lineNumber)
        {
            text = txt;
            this.startStringIndex = startStringIndex;

            this.tokens = tokens;
            this.startTokenIndex = startTokenIndex;

            this.lineNumber = lineNumber;

            var graphemes = txt.Graphemes();
            leftCorrections = new int[txt.Length];
            rightCorrections = new int[txt.Length];

            var x = 0;
            foreach (var grapheme in graphemes)
            {
                leftCorrections[x] = 0;
                rightCorrections[x] = 0;
                for (var i = 1; i < grapheme.Length; ++i)
                {
                    leftCorrections[x + i] = -i;
                    rightCorrections[x + i] = grapheme.Length - i;
                }
                x += grapheme.Length;
            }
        }

        public void adjust(Cursor cursor, int dir)
        {
            var correction = dir == -1
                ? leftCorrections
                : rightCorrections;

            if (cursor.x < correction.Length)
            {
                var delta = correction[cursor.x];
                cursor.x += delta;
                cursor.i += delta;
            }
            else if (dir == 1
                && text[text.Length - 1] == '\n')
            {
                this.adjust(cursor, -1);
            }
        }

        public string substring(int x)
        {
            return text.Substring(x);
        }

        public string substring(int x, int y)
        {
            return text.Substring(x, y);
        }

        public override string ToString()
        {
            return text;
        }
    }
}

File: js/package/src/row.js
export class Row {
    constructor(txt, tokens, startStringIndex, startTokenIndex, lineNumber) {
        this.text = txt;
        this.startStringIndex = startStringIndex;

        this.tokens = tokens;
        this.startTokenIndex = startTokenIndex;

        this.lineNumber = lineNumber;

        const graphemes = Object.freeze([...txt]),
            leftCorrections = new Array(txt.length),
            rightCorrections = new Array(txt.length);

        let x = 0;
        for (let grapheme of graphemes) {
            leftCorrections[x] = 0;
            rightCorrections[x] = 0;
            for (let i = 1; i < grapheme.length; ++i) {
                leftCorrections[x + i] = -i;
                rightCorrections[x + i] = grapheme.length - i;
            }
            x += grapheme.length;
        }

        this.adjust = (cursor, dir) => {
            const correction = dir === -1
                ? leftCorrections
                : rightCorrections;

            if (cursor.x < correction.length) {
                const delta = correction[cursor.x];
                cursor.x += delta;
                cursor.i += delta;
            }
            else if (dir === 1
                && txt[txt.length - 1] === '\n') {
                this.adjust(cursor, -1);
            }
        };

        this.toString = () => txt;

        this.substring = (x, y) => txt.substring(x, y);

        Object.seal(this);
    }

    get stringLength() {
        return this.text.length;
    }

    get endStringIndex() {
        return this.startStringIndex + this.stringLength;
    }

    get numTokens() {
        return this.tokens.length;
    }

    get endTokenIndex() {
        return this.startTokenIndex + this.numTokens;
    }
}

Row.emptyRow = (startStringIndex, startTokenIndex, lineNumber) => new Row("", [], startStringIndex, startTokenIndex, lineNumber);